src/devices/machine/sis630_gui.cpp

```diff
@@ -0,0 +1,701 @@
+	if (ACCESSING_BITS_16_23)
+		result |= downcast<sis630_svga_device *>(m_svga.target())->port_03c0_r(offset * 4 + 2) << 16;
+	if (ACCESSING_BITS_24_31)
+		result |= downcast<sis630_svga_device *>(m_svga.target())->port_03c0_r(offset * 4 + 3) << 24;
+	return result;
+}
+
+void sis630_gui_device::vga_3c0_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	if (ACCESSING_BITS_0_7)
+		downcast<sis630_svga_device *>(m_svga.target())->port_03c0_w(offset * 4 + 0, data >> 0);
+	if (ACCESSING_BITS_8_15)
+		downcast<sis630_svga_device *>(m_svga.target())->port_03c0_w(offset * 4 + 1, data >> 8);
+	if (ACCESSING_BITS_16_23)
+		downcast<sis630_svga_device *>(m_svga.target())->port_03c0_w(offset * 4 + 2, data >> 16);
+	if (ACCESSING_BITS_24_31)
+		downcast<sis630_svga_device *>(m_svga.target())->port_03c0_w(offset * 4 + 3, data >> 24);
+}
+
+u32 sis630_gui_device::vga_3d0_r(offs_t offset, uint32_t mem_mask)
+{
+	uint32_t result = 0;
+	if (ACCESSING_BITS_0_7)
+		result |= downcast<sis630_svga_device *>(m_svga.target())->port_03d0_r(offset * 4 + 0) << 0;
+	if (ACCESSING_BITS_8_15)
+		result |= downcast<sis630_svga_device *>(m_svga.target())->port_03d0_r(offset * 4 + 1) << 8;
+	if (ACCESSING_BITS_16_23)
+		result |= downcast<sis630_svga_device *>(m_svga.target())->port_03d0_r(offset * 4 + 2) << 16;
+	if (ACCESSING_BITS_24_31)
+		result |= downcast<sis630_svga_device *>(m_svga.target())->port_03d0_r(offset * 4 + 3) << 24;
+	return result;
+}
+
+void sis630_gui_device::vga_3d0_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	if (ACCESSING_BITS_0_7)
+		downcast<sis630_svga_device *>(m_svga.target())->port_03d0_w(offset * 4 + 0, data >> 0);
+	if (ACCESSING_BITS_8_15)
+		downcast<sis630_svga_device *>(m_svga.target())->port_03d0_w(offset * 4 + 1, data >> 8);
+	if (ACCESSING_BITS_16_23)
+		downcast<sis630_svga_device *>(m_svga.target())->port_03d0_w(offset * 4 + 2, data >> 16);
+	if (ACCESSING_BITS_24_31)
+		downcast<sis630_svga_device *>(m_svga.target())->port_03d0_w(offset * 4 + 3, data >> 24);
+}
+
+/*****************************
+ *
+ * 630 bridge PCI implementation
+ *
+ ****************************/
+
+DEFINE_DEVICE_TYPE(SIS630_BRIDGE, sis630_bridge_device, "sis630_bridge", "SiS 630 Virtual PCI-to-PCI bridge")
+
+sis630_bridge_device::sis630_bridge_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: pci_bridge_device(mconfig, SIS630_BRIDGE, tag, owner, clock)
+	, m_vga(*this, finder_base::DUMMY_TAG)
+{
+
+}
+
+void sis630_bridge_device::map_extra(
+	uint64_t memory_window_start, uint64_t memory_window_end, uint64_t memory_offset, address_space *memory_space,
+	uint64_t io_window_start, uint64_t io_window_end, uint64_t io_offset, address_space *io_space
+)
+{
+	if (BIT(bridge_control, 3))
+	{
+		LOGMAP("- Enable VGA control\n");
+		memory_space->install_device(0, 0xfffff, *m_vga, &sis630_gui_device::legacy_memory_map);
+		io_space->install_device(0, 0x0fff, *m_vga, &sis630_gui_device::legacy_io_map);
```

Comment:
Should these take `io_window_start` into consideration?

Reply:
Doc just mentions being installed to the legacy VGA memory/ports in this case. May effectively follow PCI bridge rules but I bet if that becomes a thing then the integration is disabled anyway. I’ll add logging at worst.